lib/charms/mysql/v0/async_replication.py

```diff
@@ -0,0 +1,777 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""MySQL async replication module."""
+
+import enum
+import logging
+import typing
+import uuid
+from functools import cached_property
+
+from charms.mysql.v0.mysql import (
+    MySQLFencingWritesError,
+    MySQLPromoteClusterToPrimaryError,
+    MySQLRejoinClusterError,
+)
+from ops import (
+    ActionEvent,
+    ActiveStatus,
+    BlockedStatus,
+    MaintenanceStatus,
+    Relation,
+    RelationDataContent,
+    Secret,
+    SecretNotFoundError,
+    WaitingStatus,
+)
+from ops.framework import Object
+from tenacity import RetryError, Retrying, stop_after_attempt, wait_fixed
+from typing_extensions import Optional
+
+from constants import (
+    BACKUPS_PASSWORD_KEY,
+    BACKUPS_USERNAME,
+    CLUSTER_ADMIN_PASSWORD_KEY,
+    CLUSTER_ADMIN_USERNAME,
+    MONITORING_PASSWORD_KEY,
+    MONITORING_USERNAME,
+    ROOT_PASSWORD_KEY,
+    ROOT_USERNAME,
+    SERVER_CONFIG_PASSWORD_KEY,
+    SERVER_CONFIG_USERNAME,
+)
+
+if typing.TYPE_CHECKING:
+    from charm import MySQLOperatorCharm
+
+logger = logging.getLogger(__name__)
+
+# The unique Charmhub library identifier, never change it
+LIBID = "4de21f1a022c4e2c87ac8e672ec16f6a"
+LIBAPI = 0
+LIBPATCH = 1
+
+PRIMARY_RELATION = "async-primary"
+REPLICA_RELATION = "async-replica"
```

Comment:
I think that we can possibly improve naming here: 
Proposal: primary instance and secondary instance only for nodes within a cluster. and primary cluster and standby cluster
Thoughts? The MySQL docs dont have naming for what I am calling a standby cluster, but I think this will help reduce confusion when discussing
We are already using standby_cluster [here](https://github.com/canonical/mysql-operator/pull/375/files#diff-bcc2205ccee68c6651d28665ca12d28056e5a7758a0dd3f7fbeab2135be51585R131)

Reply:
I would advise to keep as is, or `s/standby/replica/` where standby is present.
I said that due to mysql actually internally utilizing `replica` as a cluster role. E.g. from cluster-set-status output:

```yaml
  clusters:
    lisbon:
      clusterrole: replica
      clustersetreplicationstatus: ok
      globalstatus: ok
    rome:
      clusterrole: primary
      globalstatus: ok
      primary: db1-1.db1-endpoints.az1.svc.cluster.local:3306
  domainname: cluster-set-529e9775dd7dc7cc9c57a1b5d421bdc9
  globalprimaryinstance: db1-1.db1-endpoints.az1.svc.cluster.local:3306
  primarycluster: rome
  status: healthy
  statustext: all clusters available.
```

Reply:
not doing it now though. Have to confirm with Mohamed